app/models/allocation.rb

```diff
@@ -35,14 +35,28 @@ class  MissingNodeError < RuntimeError; end
   attr_reader :job
 
   def self.from_serialized_hash(hash)
-    job = FlightScheduler.app.job_registry.lookup(hash['job_id'])
-    return if job.nil?
     new(
-      job: job,
+      job: FlightScheduler.app.job_registry.lookup(hash['job_id']),
       nodes: hash['node_names'],
     )
   end
 
+  validate do
+    if nodes.is_a? Array && nodes.any?
+      errors.add :nodes, 'Must only contain nodes' unless nodes.all? { |n| n.is_a? Node }
```

Comment:
As a general rule I don't like using `any?` as a substitute for `! empty?`.  They don't behave the same and it always leaves me wondering if the difference is important.

```
irb(main):028:0* [].any? == ! [].empty?
=> true
irb(main):029:0> [nil].any? == ! [nil].empty?
=> false
```

---

We could combine the `any?` (`! empty?`) check with the `all?` check to have one less conditional to read.

```
if nodes.is_a? Array && ! nodes.all? { |n| n.is_a? Node }
```

Reply:
This conditional has changed quite a bit since your review:
https://github.com/openflighthpc/flight-scheduler-controller/pull/48/commits/85214b58f53741d7297b034b3e2a46bc5bedd7b4